src/Func.cc

```diff
@@ -568,24 +568,62 @@ void ScriptFunc::AddBody(StmtPtr new_body, const std::vector<IDPtr>& new_inits,
 }
 
 void ScriptFunc::ReplaceBody(const StmtPtr& old_body, StmtPtr new_body) {
-    bool found_it = false;
-
-    for ( auto body = bodies.begin(); body != bodies.end(); ++body )
-        if ( body->stmts.get() == old_body.get() ) {
-            if ( new_body ) {
-                body->stmts = new_body;
-                current_priority = body->priority;
+    if ( ! alt_body.empty() && alt_body[0].stmts == old_body ) {
+        ASSERT(new_body);
+        alt_body[0].stmts = new_body;
+    }
+    else
+        for ( auto body = bodies.begin(); body != bodies.end(); ++body )
+            if ( body->stmts == old_body ) {
+                if ( new_body ) {
+                    body->stmts = new_body;
+                    current_priority = body->priority;
+                }
+                else
+                    bodies.erase(body);
+                break;
             }
-            else
-                bodies.erase(body);
-
-            found_it = true;
-            break;
-        }
 
     current_body = new_body;
 }
 
+void ScriptFunc::SetAlternativeBody(detail::StmtPtr body, detail::ScopePtr new_scope, size_t new_frame_size) {
+    ASSERT(alt_body.empty());
+
+    Body new_body;
+    new_body.stmts = std::move(body);
+    // Priority doesn't matter because there's only one body, but let's be tidy.
+    new_body.priority = 0;
+
+    alt_body.push_back(std::move(new_body));
+    alt_scope = std::move(new_scope);
+    alt_frame_size = new_frame_size;
+
+    SwitchToAlternative();
+}
+
+void ScriptFunc::SwitchToRegular() {
+    curr_bodies = &bodies;
+    curr_frame_size = frame_size;
+    curr_scope = scope;
+}
+
+void ScriptFunc::SwitchToAlternative() {
+    curr_bodies = &alt_body;
+    curr_frame_size = alt_frame_size;
+    curr_scope = alt_scope;
+}
+
+void ScriptFunc::EnablementChanged(bool all_enabled) {
+    if ( alt_body.empty() )
+        return;
+
+    if ( all_enabled )
+        SwitchToAlternative();
+    else
+        SwitchToRegular();
+}
+
```

Comment:
Instead of adding the "alternative bodies" and the state management in Func/ScriptFunc, could you see an approach where the EventHandler's complete FuncPtr is replaced with a script_opt specific Func subclass that either executes the inlined body, or delegates execution to the original function?

Sketch below, it compiles, but nothing tested :-) Should be enough to show the idea. It does need this new `HasEnabledAllBodies()` helper and the corresponding state set during `UpdateFuncBodies()`, but that seems fine and less intrusive and possibly useful elsewhere, too.

Probably some extra-overhead due to the indirection, but keeps the Func/ScriptFunc parts more tidy and doubtful the extra indirection matters significantly. I understand this is all just because of EventGroups :smiling_face_with_tear:

```
diff --git a/src/Func.h b/src/Func.h
index d0026e245..63c3e0f21 100644
--- a/src/Func.h
+++ b/src/Func.h
@@ -84,6 +84,8 @@ public:
      */
     bool HasEnabledBodies() const { return ! bodies.empty() && has_enabled_bodies; };
 
+    bool HasEnabledAllBodies() const { return ! bodies.empty() && has_enabled_all_bodies; };
+
     /**
      * Called when there's been a change in which bodies are enabled.
      * @param all_enabled  true if now all bodies are enabled, false otherwise.
@@ -176,6 +178,7 @@ private:
     // expose accessors in the zeek:: public interface.
     friend class EventGroup;
     bool has_enabled_bodies = true;
+    bool has_enabled_all_bodies = true;
 };
 
 namespace detail {
diff --git a/src/script_opt/Inline.cc b/src/script_opt/Inline.cc
index 27e05b1c6..51971961e 100644
--- a/src/script_opt/Inline.cc
+++ b/src/script_opt/Inline.cc
@@ -245,8 +245,11 @@ void Inliner::CoalesceEventHandlers(ScriptFuncPtr func, const std::vector<Func::
 
     PostInline(oi, func, true);
 
-    new_scope = func->GetScope();
-    func->SetAlternativeBody(merged_body, new_scope, func->FrameSize());
+    auto inlined_func = make_intrusive<AlternateScriptFunc>(merged_body, new_scope, func->FrameSize(), func);
+    // Replace the function for that EventHandler the delegating one.
+    auto* eh = zeek::event_registry->Lookup(func->Name());
+    eh->SetFunc(inlined_func);
+
     funcs.emplace_back(func, new_scope, merged_body, 0);
 
     auto pf = std::make_shared<ProfileFunc>(func.get(), merged_body, true);
diff --git a/src/script_opt/ScriptOpt.h b/src/script_opt/ScriptOpt.h
index 204993885..0360207fc 100644
--- a/src/script_opt/ScriptOpt.h
+++ b/src/script_opt/ScriptOpt.h
@@ -174,6 +174,32 @@ protected:
     bool is_alt = false;
 };
 
+class StmtList;
+using StmtListPtr = zeek::IntrusivePtr<StmtList>;
+
+// ScriptFunc subclass running a single inlined body if possible,
+// otherwise delegates to the original function with multiple bodies
+class AlternateScriptFunc : public ScriptFunc {
+public:
+    AlternateScriptFunc(zeek::detail::StmtListPtr merged_body, ScopePtr scope, int frame_size, ScriptFuncPtr orig_func)
+        : orig_func(orig_func) {
+        AddBody(merged_body, frame_size);
+        SetScope(scope);
+    };
+
+    ValPtr Invoke(zeek::Args* args, Frame* parent) const override {
+        // If the original function has all bodies enabled, run our
+        // inline one, otherwise delegate
+        if ( orig_func->HasEnabledAllBodies() )
+            return ScriptFunc::Invoke(args, parent);
+
+        return orig_func->Invoke(args, parent);
+    }
+
+private:
+    ScriptFuncPtr orig_func;
+};
+

```

Reply:
> Instead of adding the "alternative bodies" and the state management in Func/ScriptFunc, could you see an approach where the EventHandler's complete FuncPtr is replaced with a script_opt specific Func subclass

Yeah, that's cleaner. I'll put together a new PR with that approach, thanks for flagging.